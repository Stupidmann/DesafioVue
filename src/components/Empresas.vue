<template>
  <b-container>
    <div class="header">
      <h1>{{title}}</h1>
      <input type="text" v-model="search" placeholder="Buscar Empresas" />
      <button @click="sortHighest">Most Rated</button>
      <div v-if="checkedNames.length > 0">
        <span>Checked: {{ checkedNames }}</span>
        <button id="favButton" @click="addToFav(checkedNames)">Add to Favorites</button>
      </div>
    </div>
    <h4>Favorites: {{favs.length}}</h4>

    <paginate name="filteredItems" :list="filteredItems" :per="3">
      <div v-for="list in filteredItems" :key="list.id" class="justify-content-md-center">

        <h2>{{list.name | removeSpecialChars | capitalize}}</h2>
        <p>{{list.description}}</p>
        <p>{{list.value}}</p>
        <p>{{list.link}}</p>
        <p>{{list.id}}</p>

        <input type="checkbox" :value='list.name' v-model="checkedNames">
        <label for=list.name>{{list.name}}:</label>

        <button id="favButton" v-on:click="addToFav(list.name)">FAV</button>

      </div>
    </paginate>

    <paginate-links for="filteredItems" :classes="{'ul': 'pagination', 'li': 'page-item','a': 'page-link'}" class="justify-content-md-center"></paginate-links>

  </b-container>
</template>

<script>
export default {

  created: function () {
    this.getEmpresas()
  },

  data () {
    return {
      checkedNames: [],
      title: 'Empresas',
      empresas: [],
      order: [],
      search: '',
      favs: [],
      paginate: ['filteredItems']
    }
  },
  methods: {
    getEmpresas () {
      const axios = require('axios')
      const url = './../challengeOne.json'
      // this.$http.get(url)
      return axios.get(url)
        .then((res) => {
          console.log(res.data.empresas)
          //  this.empresas = (res.data.empresas)
          //  data = (res.data.empresas)
          this.empresas = (res.data.empresas)
        })
        .catch(function (error) {
          console.log('Error: ', error)
        })
    },
    sortHighest () {
      this.empresas.sort((a, b) => parseInt(a.value) < parseInt(b.value) ? 1 : -1)
    },
    addToFav (id) {
      this.favs = this.favs.concat(id)
      this.checkedNames = []
      console.log(this.favs)
    }

  },
  computed: {
    filteredItems () {
      return this.empresas.filter(item => {
        return item.name.toLowerCase().indexOf(this.search.toLowerCase()) > -1
      })
    }
  },
  filters: {
    capitalize (val) {
      if (!val) return ''
      val = val.toString()
      return val.charAt(0).toUpperCase() + val.slice(1)
    },
    cleanString (val) {
      if (!val) return ''
      val = val.toString()
      return val
    },
    removeAccents (val) {
      return val
        .replace(/á/g, 'a')
        .replace(/é/g, 'e')
        .replace(/í/g, 'i')
        .replace(/ó/g, 'o')
        .replace(/ú/g, 'u')
    },
    removeSpecialChars (str) {
      return str.replace(/&\/\\#,+()$~%'":*?<>{}|]/g, '') // [ antes del '&' eliminado para hacer el commit
        .replace(/\s+/g, ' ')
        .replace(/^(\s*)([\W\w]*)(\b\s*$)/g, '$2')
    }
  }
}
</script>

<style>
    .header {
      display: flex;
      justify-content: space-between;
      align-items: center
    }
</style>
