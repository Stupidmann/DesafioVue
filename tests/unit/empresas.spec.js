import Vue from 'vue'

import { mount, shallowMount } from '@vue/test-utils'
import Empresas from '@/components/Empresas.vue'
import VuePaginate from 'vue-paginate'
import VueResource from 'vue-resource'
import BootstrapVue from 'bootstrap-vue'
import flushPromises from 'flush-promises'

Vue.use(VueResource)
Vue.use(VuePaginate)
Vue.use(BootstrapVue)

describe('component has a name', () => {
  it('renders correctly', () => {
    const wrapper = mount(Empresas)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('check title and h1', () => {
    const wrapper = mount(Empresas)
    expect(wrapper.vm.title).toBe('Empresas')
    expect(wrapper.find('h1').text()).toMatch(wrapper.vm.title)
  })

  it('calls the method when clicked', () => {
    /*  const wrapper = shallow(Empresas)
    const saveMock = jest.fn()
    wrapper.vm.save = saveMock
    wrapper.update()

    wrapper.find('#favButton')[0].trigger('click')
    expect(saveMock.mock.calls.length).toBe(1)  */
    /*  const addToFav = jest.fn()
    const wrapper = mount(Empresas, {
      methods: {
        addToFav
      }
    })
    expect(wrapper.contains('button')).toBe(true) //  contains('#favButton') devuelve falso, no encuentra el id
    const button = wrapper.find('button')
    button.trigger('click')
    expect(addToFav).toHaveBeenCalled() */
    const wrapper = shallowMount(Empresas)
    const addToFav = jest.fn()
    wrapper.setMethods({ method: addToFav })
    wrapper.find('button').trigger('click')
    expect(addToFav).toHaveBeenCalled() //  Expected mock function to have been called, but it was not called.
  })

  it('gets data from challengeOne json', () => {
    //  const wrapper = mount(Empresas) funciona aunque no se bien como
    const empresas = mount(Empresas, {
      mocks: {
        $http: {
          get: () => new Promise(resolve => resolve({ body: ['review A', 'review B'] }))
        }
      }
    })
    flushPromises().then(() => {
      expect(empresas.vm.reviews).to.deep.equal(['review A', 'review B'])
    })
  })

  it('sort highest works', () => {
    const wrapper = mount(Empresas)
    wrapper.setData({
      empresas: [{
        name: 'empresa 1',
        value: 120000
      }, {
        name: 'empresa 2',
        value: 120001
      }]
    })
    /*  expect(wrapper.vm.empresas[0]).toEqual({  //comprobar que la funcion sea llamada
      name: 'empresa 2',
      value: 120001
    })  */
    wrapper.vm.sortHighest()
    expect(wrapper.vm.empresas[0]).toEqual({
      name: 'empresa 2',
      value: 120001
    })
  })
})
