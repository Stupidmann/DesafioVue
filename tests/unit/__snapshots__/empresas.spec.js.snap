// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component has a name renders correctly 1`] = `
<div
  class="container"
>
  <div
    class="header"
  >
    <h1>
      Empresas
    </h1>
     
    <input
      placeholder="Buscar Empresas"
      type="text"
    />
     
    <button>
      Most Rated
    </button>
     
    <!---->
  </div>
   
  <h4>
    Favorites: 0
  </h4>
   
  <ul />
   
  <ul
    class="justify-content-md-center paginate-links filteredItems"
  />
</div>
`;
